### PR TITLE
Add API for raw touch events (see #17077)

### PR DESCRIPTION
**Pull request by e9925248**
First draft. OS X not even build tested

### CODE REVIEW

File: include/wx/event.h [hunk 1]
```diff
@@ -1863,31 +1872,92 @@ class WXDLLIMPEXP_CORE wxSetCursorEvent : public wxEvent
...
+
+};
+
+// MultiTouch Event
+
+class wxTouchPointId : public wxItemId<void*>
+{
+public:
+    wxTouchPointId() : wxItemId<void*>() { }
+    wxEXPLICIT wxTouchPointId(void* pItem) : wxItemId<void*>(pItem) { }
```
Comment (vadz) [inline include/wx/event.h]: It's ok to use just `explicit` in the new code, we don't support any compilers not supporting this keyword any more.

File: interface/wx/event.h [hunk 2]
```diff
@@ -3644,7 +3644,127 @@ class wxScrollEvent : public wxCommandEvent
...
+        center of zoom for Zoom event, center of rotation for Rotate event, center of box formed by 2 fingers
+        for Two Finger Tap event and position of the pressed finger for Press and Tap Event.
+    */
+    const wxPoint& GetPosition() const;
+
+    /**
+        Sets the position where the event took effect, in client coordinates: position of Pan event,
+        center of zoom for Zoom event, center of rotation for Rotate event.
+    */
+    void SetPosition(const wxPoint& pos);
```
Comment (vadz) [inline interface/wx/event.h]: I don't think this one really needs to be documented, it's for use by wxWidgets only, isn't it?

File: include/wx/event.h [hunk 3]
```diff
@@ -1863,31 +1872,92 @@ class WXDLLIMPEXP_CORE wxSetCursorEvent : public wxEvent
...
+
+    wxMultiTouchEvent(const wxMultiTouchEvent& event) : wxTouchEventBase(event)
+    {
+        m_isPrimary = event.m_isPrimary;
+        m_sequence = event.m_sequence;
+    }
+
+    bool IsPrimary() const { return m_isPrimary; }
+    void SetPrimary(bool primary) { m_isPrimary = primary; }
+    const wxTouchPointId& GetSequence() const { return m_sequence; }
```
Comment (vadz) [inline include/wx/event.h]: This is a bit confusing, why is "sequence" an "ID"? Shouldn't a sequence consist of several IDs? If not, I suggest renaming `wxTouchPointId` to `wxTouchSequenceId' although, to be honest, I'm not sure if it's appropriate to call it an ID at all, considering that it doesn't seem to be used to identify anything in the same sense as other `wxItemID`-derived classes do.

File: samples/touchtest/touchtest.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+// Licence:     wxWindows licence
+/////////////////////////////////////////////////////////////////////////////
+
+#include "wx/wx.h"
+
+// Define a new application
+class MyApp: public wxApp
+{
+public:
+    bool OnInit();
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: Just to give a good example, could you please add `wxOVERRIDE` here?

File: samples/touchtest/touchtest.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+};
+
+DECLARE_APP(MyApp)
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: Similarly, please use `wx` prefix and add a semicolon at the end of line here.

File: samples/touchtest/touchtest.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+#define TOUCH_POINTS 16
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: Could we please use `const unsigned TOUCH_POINTS = 16;` here instead?

File: samples/touchtest/touchtest.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+class MyFrame: public wxFrame
+{
+private:
+    typedef struct
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: There is no need for this `typedef` in C++, please just use `struct TouchState`.

File: samples/touchtest/touchtest.cpp [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    m_TouchPoints[10].pen = wxPen(*wxGREEN, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[11].pen = wxPen(*wxRED, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[12].pen = wxPen(*wxBLACK, 2, wxPENSTYLE_DOT);
+    m_TouchPoints[13].pen = wxPen(*wxYELLOW, 2, wxPENSTYLE_DOT);
+    m_TouchPoints[14].pen = wxPen(*wxBLUE, 2, wxPENSTYLE_DOT);
+    m_TouchPoints[15].pen = wxPen(*wxRED, 2, wxPENSTYLE_DOT);
+
+    wxMenu *file_menu = new wxMenu;
+
+    file_menu->Append(wxID_EXIT, wxT("&Exit"));
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: There is no need to specify the label for the stock IDs, doing it just prevents the standard platform accelerator for it from working, so it's better to just remove it.

File: samples/touchtest/touchtest.cpp [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+    wxMenuBar *menu_bar = new wxMenuBar;
+
+    menu_bar->Append(file_menu, wxT("&File"));
+    SetMenuBar(menu_bar);
+
+    EnableTouchEvents(wxTOUCH_RAW_EVENTS);
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: It would probably be useful to test for the return value here and give some error message if it failed.

File: samples/touchtest/touchtest.cpp [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+void MyFrame::OnQuit(wxCommandEvent& WXUNUSED(event))
+{
+    Close(true);
+}
+
+void MyFrame::DrawStart(const wxTouchPointId& id, wxPoint pos)
+{
+    int idx = -1;
+    for(unsigned i = 0; i < TOUCH_POINTS; i++)
+        if (m_TouchPoints[i].id == id)
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: Could you please put braces around composite statements bodies taking more than one line? It would make them much more readable (here and below).

File: samples/touchtest/touchtest.cpp [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	    }
+    if (idx == -1)
+        return;
+    m_TouchPoints[idx].id = id;
+    m_TouchPoints[idx].last = pos;
+}
+
+void MyFrame::DrawUpdate(const wxTouchPointId& id, wxPoint pos)
+{
+    int idx = -1;
```
Comment (vadz) [inline samples/touchtest/touchtest.cpp]: This repeated snippet of code would become more clear if it were extracted into some `FindIndexOfTouchId()` function.

File: src/gtk/window.cpp [hunk 12]
```diff
@@ -3302,6 +3303,41 @@ touch_callback(GtkWidget* WXUNUSED(widget), GdkEventTouch* gdk_event, wxWindowGT
...
+        case GDK_TOUCH_END:
+            type = wxEVT_TOUCH_END;
+            break;
+
+        case GDK_TOUCH_CANCEL:
+            type = wxEVT_TOUCH_CANCEL;
+            break;
+
+	default:
+	    type = wxEVT_NULL;
```
Comment (vadz) [inline src/gtk/window.cpp]: Something wrong with the indentation in these 2 lines.

File: src/gtk/window.cpp [hunk 13]
```diff
@@ -3302,6 +3303,41 @@ touch_callback(GtkWidget* WXUNUSED(widget), GdkEventTouch* gdk_event, wxWindowGT
...
+        }
+	if (type != wxEVT_NULL)
+	{
+            wxMultiTouchEvent event(type);
+	    event.SetPosition(wxPoint(gdk_event->x, gdk_event->y));
```
Comment (vadz) [inline src/gtk/window.cpp]: And here as well. I won't repeat this comment any more, but please check that you don't have any hard TABs in your changes and replace them with spaces if you do. TIA!

File: src/gtk/window.cpp [hunk 14]
```diff
@@ -3302,6 +3303,41 @@ touch_callback(GtkWidget* WXUNUSED(widget), GdkEventTouch* gdk_event, wxWindowGT
...
+	    event.SetSequence(wxTouchPointId(gdk_event->sequence));
+	    event.SetPrimary(gdk_event->emulating_pointer);
+            win->HandleWindowEvent(event);
```
Comment (vadz) [inline src/gtk/window.cpp]: Should we really continue with the other events generation if this one was handled?

File: src/msw/window.cpp [hunk 15]
```diff
@@ -919,6 +952,17 @@ bool wxWindowMSW::EnableTouchEvents(int eventsMask)
         // dynamically.
         wxVector<GESTURECONFIG> configs;
 
+        if ( eventsMask & wxTOUCH_RAW_EVENTS )
+        {
+            eventsMask = eventsMask & ~wxTOUCH_RAW_EVENTS;
```
Comment (vadz) [inline src/msw/window.cpp]: `&=`?

File: src/msw/window.cpp [hunk 16]
```diff
@@ -3382,6 +3426,17 @@ wxWindowMSW::MSWHandleMessage(WXLRESULT *result,
...
         }
         break;
+
+        case WM_TOUCH:
+        {
+            if ( !GestureFuncs::IsOk() )
+                break;
+
+            HandleTouch(wParam, lParam);
+            // Let DefWindowProc handle the touch events too
```
Comment (vadz) [inline src/msw/window.cpp]: Again, IMO we should only let unprocessed events through.

File: src/msw/window.cpp [hunk 17]
```diff
@@ -5860,6 +5915,36 @@ bool wxWindowMSW::HandlePressAndTap(const wxPoint& pt, WXDWORD flags)
 
     return HandleWindowEvent(event);
 }
+
+bool wxWindowMSW::HandleTouch(WXWPARAM wParam, WXLPARAM lParam)
+{
+    const unsigned count = LOWORD(wParam);
+    TOUCHINPUT info[count];
```
Comment (vadz) [inline src/msw/window.cpp]: This is not standard C++, we need to use `wxVector` here to allocate an array of dynamic size.

File: src/msw/window.cpp [hunk 18]
```diff
@@ -5860,6 +5915,36 @@ bool wxWindowMSW::HandlePressAndTap(const wxPoint& pt, WXDWORD flags)
...
+    if (GestureFuncs::GetTouchInputInfo() ((HTOUCHINPUT)lParam, count, info, sizeof(TOUCHINPUT)))
+    {
+        for(unsigned i = 0; i < count; i++)
+        {
+            wxPoint pos = ScreenToClient(wxPoint(info[i].x / 100, info[i].y / 100));
```
Comment (vadz) [inline src/msw/window.cpp]: A comment explaining these factors of 100 would be nice.

File: src/osx/cocoa/window.mm [hunk 19]
```diff
@@ -23,6 +23,7 @@
     #include "wx/osx/private.h"
 #endif
 
+#include "wx/dynarray.h"
```
Comment (vadz) [inline src/osx/cocoa/window.mm]: Please don't use this in the new code, use `wxVector<>`.

File: src/osx/cocoa/window.mm [hunk 20]
```diff
@@ -2050,6 +2082,84 @@ When dragging the bottom part of the DND sample ("Drag text from here!")
...
+	if (other == NULL)
+            continue;
+        if ([id isEqual:other])
+	{
+	    if (type == wxEVT_TOUCH_END || type == wxEVT_TOUCH_CANCEL)
+	    {
+                [other release];
+                m_touchIDs[i] = NULL;
+            }
+	    return (void*)(intptr_t)i;
```
Comment (vadz) [inline src/osx/cocoa/window.mm]: We have `wxUIntToPtr()` for this kind of casts.

**Review by vadz — COMMENTED** [review]
Thanks for your changes!

Globally they look fine to me, there are just some small things here and there that it would be nice (and hopefully not difficult) to improve, but I have one big question about the meaning of touch IDs/sequence as I just don't understand what do they actually represent. Perhaps it would be clear to somebody familiar with the platform touch input IDs, but I'm just lost here and the example of using them in the sample unfortunately doesn't help a lot as I don't understand what is it supposed to do. Could you please add some high(er) level explanation of this to the documentation and maybe some comments explaining what's going on to the sample?

TIA!

File: samples/touchtest/touchtest.cpp (hunk 6 above)
Comment (e9925248) [reply samples/touchtest/touchtest.cpp]: Only constexpr would fit - see m_TouchPoints array

File: include/wx/event.h (hunk 3 above)
Comment (e9925248) [reply include/wx/event.h]: The ID tracks the individual fingers: Each finger can be compared to a individual cursor - in the event you need to know, which one has moved. As various plattforms have got different ways for implementing the ID, it is an opaque type just supporting copying and checking equalitity.

File: src/msw/window.cpp (hunk 16 above)
Comment (e9925248) [reply src/msw/window.cpp]: DefWindowProc takes care of memory cleanup. 

In regard to platform agnostic behavior: GTK and OS X does not allow returning any handled flag for an event, so I wouldn't do it for a multi-platform solution.

File: src/msw/window.cpp (hunk 18 above)
Comment (e9925248) [reply src/msw/window.cpp]: https://msdn.microsoft.com/de-de/library/windows/desktop/dd317334(v=vs.85).aspx
This member is indicated in hundredths of a pixel of physical screen coordinates.

File: src/gtk/window.cpp (hunk 14 above)
Comment (e9925248) [reply src/gtk/window.cpp]: We can only block the the detection of some gesture ["some" = gestures detected by the wxWidgets itself, so a platform specific set].

File: include/wx/event.h (hunk 3 above)
Comment (vadz) [reply include/wx/event.h]: OK, thanks, but why is the accessor called `GetSequence()`? Also, maybe `wxTouchPointId` should be `wxTouchCursorId` then?

Comment (e9925248) [reply include/wx/event.h]: Sequence has been take from GTK. Windows calls the field dwID. OS X call it identity.
https://msdn.microsoft.com/de-de/library/windows/desktop/dd317334(v=vs.85).aspx
https://developer.apple.com/documentation/appkit/nstouch/1535399-identity

I would not call it cursor, as this is the blinking onscreen object.

Comment (vadz) [reply include/wx/event.h]: Again, my main problem is inconsistency between "point" and "sequence" which seem to be totally different to me, but are actually the same thing. I admit I don't like the use of "sequence" in this context very much neither, but I think using `wxTouchSequenceId` for the type and `GetSequenceId()` for the method would be reasonably clear.

Alternatively, we could use `wxTouchEventIdentity` and `GetEventIdentity()` if you wish, but we really shouldn't abbreviate "Identity" to "ID" as an ID already means something else in wx events context.

"Cursor" is indeed bad, I only used it because you mentioned it in the first reply, but I don't really like it neither.

To summarize: let's use "SequenceId" if you don't mind or please feel free to choose anything else but please use the same or similar names for the type and the corresponding accessor and let's try not to overload "event id" with yet another meaning.

**Comment by vadz** [comment]
Apparently some changes are needed to make this work under MSW, see [this post](https://groups.google.com/d/msg/wx-dev/JdeuH95_bGQ/tcJCXMZUDgAJ).

**Comment by vadz** [comment]
Unfortunately this won't make it into 3.2 unless somebody steps up and does all the work needed to integrate it. I wish I had time to do this myself, but I don't.

**Comment by dsa-t** [comment]
What's the current status of this?

**Comment by vadz** [comment]
Sorry, I don't think the original submitter is active any more and nobody else has come up to take over this one.

**Comment by vadz** [comment]
Replaced by #24610.
